### PR TITLE
Separate RequiredMappingStrategy configuration for enums

### LINKED ISSUE #1590: Separate RequiredMappingStrategy configuration for enums
**Is your feature request related to a problem? Please describe.**
When there are multiple properties on the Source you want to ignore because they are not present (and relevant on the target) setting RequiredMappingStrategy = RequiredMappingStrategy.Target is super convenient. However, setting this property also removes the warnings when the source has Enum values that are not present on the Target, which can cause runtime exceptions.

**Describe the solution you'd like**
I would like the RequiredMappingStrategy.Target to still give warnings when the source has enum values which are not present in the target.

**Describe alternatives you've considered**
The alternative is not using RequiredMappingStrategy.Target and ignoring the members not present on the source manually using [MapperIgnoreSource] but that is much less convenient in many cases.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by jacob-buckaroo**
# Separate RequiredMappingStrategy configuration for enums

## Description

Add MapperAttribute.RequiredEnumMappingStrategy / MapperDefaultsAttribute.RequiredEnumMappingStrategy
Resolve RequiredMappingStrategy for each enum mapping in the following order, fall back to the next one if the previous one is not set:
1. MapperRequiredMappingAttribute on the mapping method
2. MapperAttribute.RequiredEnumMappingStrategy
3. MapperDefaultsAttribute.RequiredEnumMappingStrategy
4. MapperAttribute.RequiredMappingStrategy
5. MapperDefaultsAttribute.RequiredMappingStrategy

Fixes #1590

## Checklist

- [x] The existing code style is followed
- [x] The commit message follows our guidelines
- [x] Performed a self-review of my code
- [x] Hard-to-understand areas of my code are commented
- [x] The documentation is updated (as applicable)
- [x] Unit tests are added/updated
- [x] Integration tests are added/updated (as applicable, especially if feature/bug depends on roslyn or framework version in use)

### CODE REVIEW

File: src/Riok.Mapperly.Abstractions/RequiredEnumMappingStrategy.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+namespace Riok.Mapperly.Abstractions;
+
+/// <summary>
+/// Defines the strategy used when emitting warnings for unmapped members.
+/// </summary>
+[Flags]
+public enum RequiredEnumMappingStrategy
```
[the remark this answers is not in the document]
Comment (jacob-buckaroo) [reply src/Riok.Mapperly.Abstractions/RequiredEnumMappingStrategy.cs]: Fixed.

File: src/Riok.Mapperly/Configuration/MapperConfigurationMerger.cs [hunk 2]
```diff
@@ -51,6 +51,11 @@ public static MapperAttribute Merge(MapperConfiguration mapperConfiguration, Map
             ?? defaultMapperConfiguration.RequiredMappingStrategy
             ?? mapper.RequiredMappingStrategy;
 
+        mapper.RequiredEnumMappingStrategy =
+            mapperConfiguration.RequiredEnumMappingStrategy
+            ?? defaultMapperConfiguration.RequiredEnumMappingStrategy
+            ?? mapper.RequiredEnumMappingStrategy;
```
[the remark this answers is not in the document]
Comment (jacob-buckaroo) [reply src/Riok.Mapperly/Configuration/MapperConfigurationMerger.cs]: Fixed.

File: src/Riok.Mapperly/Configuration/MapperConfigurationReader.cs [hunk 3]
```diff
@@ -164,6 +175,9 @@ private EnumMappingConfiguration BuildEnumConfig(MappingConfigurationReference c
             ignoredTargets,
             explicitMappings,
             requiredMapping ?? MapperConfiguration.Enum.RequiredMappingStrategy,
+            requiredMapping is not null
+                ? (RequiredEnumMappingStrategy)requiredMapping
+                : MapperConfiguration.Enum.RequiredEnumMappingStrategy,
```
[the remark this answers is not in the document]
Comment (jacob-buckaroo) [reply src/Riok.Mapperly/Configuration/MapperConfigurationReader.cs]: Fixed.

File: src/Riok.Mapperly/Configuration/MembersMappingConfiguration.cs [hunk 4]
```diff
@@ -10,7 +10,8 @@ public record MembersMappingConfiguration(
     IReadOnlyCollection<MemberMappingConfiguration> ExplicitMappings,
     IReadOnlyCollection<NestedMembersMappingConfiguration> NestedMappings,
     IgnoreObsoleteMembersStrategy IgnoreObsoleteMembersStrategy,
-    RequiredMappingStrategy RequiredMappingStrategy
+    RequiredMappingStrategy RequiredMappingStrategy,
+    RequiredEnumMappingStrategy RequiredEnumMappingStrategy
```
[the remark this answers is not in the document]
Comment (jacob-buckaroo) [reply src/Riok.Mapperly/Configuration/MembersMappingConfiguration.cs]: Fixed.

File: src/Riok.Mapperly/Configuration/MapperConfigurationReader.cs [hunk 5]
```diff
@@ -126,7 +136,8 @@ private MembersMappingConfiguration BuildMembersConfig(MappingConfigurationRefer
             memberConfigurations,
             nestedMembersConfigurations,
             ignoreObsolete ?? MapperConfiguration.Members.IgnoreObsoleteMembersStrategy,
-            requiredMapping ?? MapperConfiguration.Members.RequiredMappingStrategy
+            requiredMapping ?? MapperConfiguration.Members.RequiredMappingStrategy,
+            MapperConfiguration.Members.RequiredEnumMappingStrategy
```
[the remark this answers is not in the document]
Comment (jacob-buckaroo) [reply src/Riok.Mapperly/Configuration/MapperConfigurationReader.cs]: Fixed.

File: src/Riok.Mapperly/Configuration/EnumMappingConfiguration.cs [hunk 6]
```diff
@@ -11,6 +11,7 @@ public record EnumMappingConfiguration(
     IReadOnlyCollection<IFieldSymbol> IgnoredTargetMembers,
     IReadOnlyCollection<EnumValueMappingConfiguration> ExplicitMappings,
     RequiredMappingStrategy RequiredMappingStrategy,
+    RequiredMappingStrategy RequiredEnumMappingStrategy,
```
Comment (latonz) [inline src/Riok.Mapperly/Configuration/EnumMappingConfiguration.cs:14]: I don't think this member is needed, is it? The existing `RequiredMappingStrategy` can be used.

File: docs/docs/configuration/enum.mdx [hunk 7]
```diff
@@ -182,18 +182,18 @@ public partial class CarMapper
 
 By default Mapperly emits diagnostics with a severity of warning if there are unmapped source or target enum members.
 To enforce strict enum member mappings on only either the source or the target,
-the `RequiredMappingStrategy` can be used.
+the `RequiredEnumMappingStrategy` can be used. By default this property is set to `Inherit`, which will inherit the value from `RequiredMappingStrategy`.
```
Comment (latonz) [inline docs/docs/configuration/enum.mdx:185]: I think the `By default ...` sentence needs to be updated as it inherits but it is not set to a member Inherit...

**Review by latonz — CHANGES_REQUESTED** [review]
Thanks for the updates, two small feedback points...

File: src/Riok.Mapperly/Configuration/EnumMappingConfiguration.cs (hunk 6 above)
Comment (jacob-buckaroo) [reply src/Riok.Mapperly/Configuration/EnumMappingConfiguration.cs:14]: When I remove this from EnumMappingConfiguration, MapperConfigurationReader & EnumMappingDiagnosticReporter both these test don't work anymore:
- MapperAttributeRequiredEnumMappingSourceAndRequiredMappingNoneWithUnmappedMember
- MapperAttributeRequiredEnumMappingTargetAndRequiredMappingNoneWithUnmappedMember

In these tests I needed to change `ctx.Configuration.Enum.RequiredEnumMappingStrategy` tp `ctx.Configuration.Enum.RequiredMappingStrategy`.

Comment (latonz) [reply src/Riok.Mapperly/Configuration/EnumMappingConfiguration.cs:14]: If I change it, all tests pass, find my changes I quickly tried in this [patch](https://github.com/user-attachments/files/17830203/patch.patch). Why don't the tests work?

File: docs/docs/configuration/enum.mdx [hunk 8]
```diff
@@ -182,18 +182,18 @@ public partial class CarMapper
...
     
-        Sets the `RequiredMappingStrategy` for all methods within the mapper,
+        Sets the `RequiredEnumMappingStrategy` for all methods within the mapper,
         by default it is `Both` requiring all members to be mapped.
         This can be overriden by individual mapping methods using `MapperRequiredMappingAttribute`.
     
         ```csharp
         // highlight-start
-        [Mapper(RequiredMappingStrategy = RequiredMappingStrategy.Source)]
+        [Mapper(RequiredEnumMappingStrategy = RequiredEnumMappingStrategy.Source)]
```
Comment (latonz) [inline docs/docs/configuration/enum.mdx:196]:
```suggestion
        [Mapper(RequiredEnumMappingStrategy = RequiredMappingStrategy.Source)]
```

**Review by latonz — CHANGES_REQUESTED** [review]
Added my feedback...

File: src/Riok.Mapperly/Configuration/EnumMappingConfiguration.cs (hunk 6 above)
Comment (jacob-buckaroo) [reply src/Riok.Mapperly/Configuration/EnumMappingConfiguration.cs]: I'm not sure what I did wrong, it indeed works as expected now. Thanks!

File: docs/docs/configuration/enum.mdx (hunk 7 above)
Comment (jacob-buckaroo) [reply docs/docs/configuration/enum.mdx]: Fixed.

**Pull request merged**